### PR TITLE
Added FeatureSelection transformer

### PR DESCRIPTION
**Pull request by aiwalter**
Added FeatureSelection transformer

#### Reference Issues/PRs
#1268 

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/alan-turing-institute/sktime/blob/main/.all-contributorsrc).
- [x] Optionally, I've updated sktime's [CODEOWNERS](https://github.com/alan-turing-institute/sktime/blob/main/CODEOWNERS) to receive notifications about future changes to these files.
- [ ] I've added unit tests and made sure they pass locally.

##### For new estimators
- [x] I've added the estimator to the online documentation.
- [ ] I've updated the existing example notebooks or provided a new one to showcase how my estimator works.

### CODE REVIEW

**Comment by aiwalter** [comment]
Todo: add method `ignore` to set `Z` to `None`

**Comment by fkiraly** [comment]
I think @ViktorKaz already has this in his PR on the #856 , so perhaps good to align?

**Comment by aiwalter** [comment]
Oh didnt see, but seem my implementation is a superset as it has more methods/algos. @ViktorKaz whats your opinion? Not sure as you implemented a `_PanelToPanelTransformer`?

**Comment by fkiraly** [comment]
oh, I see - you do automated feature selection, so it's not just a subsetter. Please ignore my comment then.

**Comment by ViktorKaz** [comment]
@fkiraly @aiwalter the transformer that I implemented is quite simple, it just selects columns based on the column indices. From what I quickly saw `sktime.transformations.series.feature_selection.FeatureSelection` can also select features by importance or at random which is nice and in this respect is better than my transformer.

I implemented a `_PanelToPanelTransformer` because I needed a transformer that takes a pandas dataframe as input and returns a pandas series. From memory, I initially tried implementing `_PanelToSeriesTransformer` (I suppose this is the correct type) but ran into some issues with the unit tests.

Based on a quick look at  `sktime.transformations.series.feature_selection.FeatureSelection` I don't understand why `Z` can be a pd.Series. If the input has only one column (feature) what is there to select?

**Comment by aiwalter** [comment]
okay, so then we can have both of this transformers if I understand it correctly. Pls feel free to propose other feature selection methods I could add to this transformer.

@ViktorKaz I accept only `pd.DataFrame`, was already in the code but wrong in docstring, updated it now. Thx!

**Comment by aiwalter** [comment]
@fkiraly I think we need another tag `multivariate-only` as equivalence to `univariate-only`? Tests are failing as the `FeatureSelection` transformer can only deal with `pd.DataFrame`. Other solution is that in case it gets a `pd.Series` it would just pass this through. What do you think?

**Comment by aiwalter** [comment]
@fkiraly I am struggling a bit with the tests as I have a line `Z = check_series(Z, enforce_multivariate=True)`, so some tests are failing here. I am wondering if that should actually be ignored then by the tests? I though tests accept some exceptions?

**Comment by fkiraly** [comment]
I didn't touch *transformers* in relation to the univariate/multivariate discussion.
Probably we need to make case distinctions, and possibly even introduce a tag for transformers that require multivariate input?

If that's the "deeper problem" here, would you mind adding this as part of the PR?

Any opinions, @mloning, @GuzalBulatova, @thayeylolu?

**Comment by aiwalter** [comment]
can I use the `scitype:y": "multivariate` to enforce multivariate? Or do we need a `multivariate-only` tag?

**Review by fkiraly — CHANGES_REQUESTED** [review]
Ok, here's my opinion on the test logic.

I think it's in general a good idea to change the tests so they run with multivariate series, and it's much needed. However, I think the way you are doing it is not good.

What I think is not a good idea:
* you are modifying the signature of all tests, to pass some side information in (`n_columns`). That's quite some heavy change there in the architecture.
* You are, in fact, passing the same information twice - once as the `scitype:y` tag (of the `Estimator` argument), and once as the `n_columns` argument. That's redundant. Why don't you parse the tags just in the sub-tests where it is needed, instead of parsing it in the "master test" and then passing it on? This is how it is done for controlling test behaviour in other cases, too.
* the tag `scitype:y` also seems confusing, because there's no `y` in the transformers? I suggest to use a tag, but not this name for it.

My proposed change to this would be:
* leave the test signatures as they were, only taking the estimator
* in the tests where you need `n_columns` to be determined by the estimator, use the tag.
* Stick to the general pattern that's currently used, if you need to control test behaviour, use the tags. I don't think there's a case where we can't do this, right?
* minor note, if you introduce a new tag or use it for an an estimator scitype it hasn't been used before, you ought to update the `_tag` registry

**Comment by aiwalter** [comment]
> Why don't you parse the tags just in the sub-tests where it is needed, instead of parsing it in the "master test" and then passing it on? This is how it is done for controlling test behavior in other cases, too.

I did this in order to avoid copy and pasting if statements like `if Estimator.get_class_tag("scitype:y") == "multivariate":`, bcz this is what I would need to do for almost any test then.

> the tag scitype:y also seems confusing, because there's no y in the transformers? I suggest to use a tag, but not this name for it.

Ok but then I need to introduce a ` scitype:Z` tag

**Comment by fkiraly** [comment]
>I did this in order to avoid copy and pasting if statements like if Estimator.get_class_tag("scitype:y") == "multivariate":, bcz this is what I would need to do for almost any test then.

In `test_all_transformers`, it's two out of nine, in `estimator_checks` it's about half, that's less than the majority, not "almost any".

But I get what you are trying to do, I would suggest one of the two:

* do it in `_get_args`, that way you don't change the general test signature but only how the args are created
* I anticipate that we may want to check multiple args, e.g., `n_columns=1` and `n_columns=2` depending on the various tags. You could do a bit of a refactor by creating a `_yield_args`, and then looping over all scenarios, though that's perhaps too much of a refactor. 

>Ok but then I need to introduce a `scitype:Z` tag

That would be fine with me.

**Comment by aiwalter** [comment]
but why `_get_args`? I guess you mean `_make_args`?

**Comment by fkiraly** [comment]
>but why `_get_args`? I guess you mean `_make_args`?

yes, that was a typo - I mean `_make_args`.

**Comment by fkiraly** [comment]
Ok, I've thought about this some more.

I think we should apply the general principle that we have in our governance docs: if there's a major refactor of APIs, there ought to be a STEP.

Here's what I think:
* changing signature of tests is a major change - who knows what this might break down the line, so it needs a description of what is changing in the tests API, and @mloning also needs to have a look at that. I would also do this in a separate PR.
* Changing the *logic* of individual tests is not a major change though, an I'd say it's fine as long as it's properly documented what, say, `_make_args` does.

File: sktime/forecasting/base/_meta.py [hunk 1]
```diff
@@ -32,7 +32,7 @@ def __init__(self, forecasters, n_jobs=None):
 
     def _check_regressor(self):
         if self.regressor is None:
-            self.regressor_ = GradientBoostingRegressor(max_depth=5)
+            self.regressor_ = GradientBoostingRegressor(max_depth=5, random_state=1)
```
Comment (mloning) [inline sktime/forecasting/base/_meta.py:35]: Shouldn't we pass a random state here which a user can define in the constructor?
```suggestion
            self.regressor_ = GradientBoostingRegressor(max_depth=5, random_state=self.random_state)
```

File: sktime/transformations/series/feature_selection.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+        # multivariate Z
+        if not isinstance(Z, pd.Series):
+            if self.method == "feature-importances":
+                self._check_regressor()
+                self._check_n_columns(Z)
+                X = check_series(X)
+                # fit regressor with Z as exog data and X as endog data (target)
+                self.regressor_.fit(X=Z, y=X)
+                # create dict with columns name (key) and feauter importance (value)
+                d = dict(zip(Z.columns, self.regressor_.feature_importances_))
```
Comment (mloning) [inline sktime/transformations/series/feature_selection.py:121]: What happens if `feature_importances_` isn't an attribute of the `self.regressor_`? I saw that you describe that in the docstring which is great but perhaps we could check for that an raise an error message?

**Review by mloning — COMMENTED** [review]
Hi @aiwalter had a quick look, just two minor comments.

File: sktime/forecasting/base/_meta.py (hunk 1 above)
Comment (aiwalter) [reply sktime/forecasting/base/_meta.py:35]: I was not sure on this, I thought if user want to set seed, a `regressor` can be given with the respective params. But let me add a param `random_state ` as a shorthand

File: sktime/transformations/series/feature_selection.py (hunk 2 above)
Comment (aiwalter) [reply sktime/transformations/series/feature_selection.py:121]: yes, let me add a check for this to raise a proper exception message

**Comment by aiwalter** [comment]
@mloning thx for review. Did some changes and would appreciate to have it in next release

**Review by mloning — DISMISSED** [review]
Thanks @aiwalter, looks good to me now.

**Comment by aiwalter** [comment]
@fkiraly pls review

File: sktime/forecasting/trend.py [hunk 3]
```diff
@@ -207,6 +207,9 @@ def _predict(self, fh=None, X=None, return_pred_int=False, alpha=DEFAULT_ALPHA):
         y_pred_int : pd.DataFrame
             Prediction intervals for the forecast
         """
+        if return_pred_int:
```
Comment (fkiraly) [inline sktime/forecasting/trend.py:210]: please revert this - it has nothing to do with the PR, and the `NotImplementedError` is dealt with by the base class

File: sktime/tests/_config.py [hunk 4]
```diff
@@ -332,6 +333,7 @@
     Imputer: {"method": "mean"},
     HampelFilter: {"window_length": 3},
     OptionalPassthrough: {"transformer": BoxCoxTransformer(), "passthrough": True},
+    FeatureSelection: {"method": "all"},
```
Comment (fkiraly) [inline sktime/tests/_config.py:336]: please add this to `get_test_params` instead, new policy on where the defaults are 😄

Comment (fkiraly) [reply sktime/tests/_config.py:336]: tests use `get_test_params` which (for downwards compatibility) looks in `_config`, but the primary location for the parameters is in `get_test_params` now

**Review by fkiraly — CHANGES_REQUESTED** [review]
I think this is great and almost ready.

But, there are some minor things which I think should be fixed before merging:

* adding superfluous tag handling in `TrendForecaster` which is out of scope and a duplication
* default test params should move in the designated method from `_config`

File: sktime/forecasting/trend.py (hunk 3 above)
Comment (aiwalter) [reply sktime/forecasting/trend.py:210]: I think I can remove this `raise` completely then, rather then reverting?

Comment (aiwalter) [reply sktime/forecasting/trend.py]: ok I think this is resolved now as I removed it completely, which seems to be the `main` state as well.

**Comment by aiwalter** [comment]
@Aparna-Sakshi I have a quick question regarding your PR #1361. My tests are failing, seems bcz I have different params in `get_test_params` than in the default `init`. But this should in my opinion be allowed? Pls correct me if this issue is not coming from the `get_test_params`

also tagging @fkiraly

**Comment by fkiraly** [comment]
I think this is perhaps a logic error in the test `check_constructor`?
Or, it is an undocumented requirement for the test instance?
Namely, the test assumes that the first test instance's parameters are equal to the defaults, otherwise it raises an error.

This logic did not change when we made the interface change - you can check by reverting to an earlier version and making the same change in `_config` .

**Comment by fkiraly** [comment]
My suggestion would be: for this PR, be compliant with the usual test logic where it is assumed that the `_config` parameters do not overwrite any defaults.

In a separate PR, we update the test logic to:
* allow overrides of default parameter values
* loop through all elements of `get_test_params` if it is a list.

**Comment by fkiraly** [comment]
@aiwalter, actually, it was a bug caused by a remaining reference to the `_config`, in the failing test.

This is now fixed in #1406, have a look.

**Comment by aiwalter** [comment]
Okay thx. Then let me try again with the new way

**Comment by fkiraly** [comment]
While you are at it:

I started refactoring the tests since I thought it might be good to have multiple parameter combinations tested per estimator, and that was not possible with the rigid, non-pytest way the tests and fixtures were looped over.

I've fixed this now and would appreciate your review of #1407.

**Comment by aiwalter** [comment]
@fkiraly should `get_test_params` not be private `_get_test_params`?

**Comment by mloning** [comment]
+1 for making them private

File: sktime/transformations/series/feature_selection.py [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+        super(FeatureSelection, self).__init__()
+
+    def fit(self, Z, X=None):
+        """Fit the transformation on input series `Z`.
+
+        Parameters
+        ----------
+        Z : pd.DataFrame
+            A time series to apply the transformation on.
+
```
Comment (mloning) [inline sktime/transformations/series/feature_selection.py:101]: docstring for X is missing

File: sktime/transformations/series/feature_selection.py [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+                    raise AttributeError("Parameter columns must be given.")
+                self.columns_ = self.columns
+            elif self.method == "none":
+                self.columns_ = None
+            elif self.method == "all":
+                self.columns_ = list(Z.columns)
+            else:
+                raise ValueError("Incorrect method given. Try another method.")
+
+        self._is_fitted = True
```
Comment (mloning) [inline sktime/transformations/series/feature_selection.py:149]: This is handled in the base class
```suggestion

```

File: sktime/transformations/series/feature_selection.py [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+        Returns
+        -------
+        self
+        """
+        Z = check_series(Z)
+        self.n_columns_ = self.n_columns
+        self.feature_importances_ = None
+
+        # multivariate Z
+        if not isinstance(Z, pd.Series):
```
Comment (mloning) [inline sktime/transformations/series/feature_selection.py:111]: for readability it may be easier to read:
```suggestion
        if isinstance(Z, pd.Series):
            return self
        else:
```

File: sktime/transformations/series/feature_selection.py [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+            else:
+                Zt = Z
+        return Zt
+
+    def _check_n_columns(self, Z):
+        if not isinstance(self.n_columns_, int):
+            self.n_columns_ = int(math.ceil(Z.shape[1] / 2))
+
+    @classmethod
+    def get_test_params(cls):
```
Comment (mloning) [inline sktime/transformations/series/feature_selection.py:187]: Shouldn't this be implemented in the base class? The test parameters should then be a class variable similar to the tags.

**Review by mloning — COMMENTED** [review]
Hi @aiwalter, I had a quick look, just some minor comments.

File: sktime/transformations/series/feature_selection.py (hunk 8 above)
Comment (aiwalter) [reply sktime/transformations/series/feature_selection.py:186]: I couldnt find how to actually implement this. It says its in the extension template but there I cant find it.

Comment (fkiraly) [reply sktime/transformations/series/feature_selection.py:186]: that's because we separated the refactor (moving logic) from adding the documentation in the extension templates.

The extension template PR is still a draft and under review here:
https://github.com/alan-turing-institute/sktime/pull/1395

The docstring should say though what it should do:
return either a dict of params or a list of dict; in either case, the dicts must be suitable for passing to the init, i.e., `__init__(**mydict)` must result in a sensible test instance for the class.

Comment (fkiraly) [reply sktime/transformations/series/feature_selection.py:186]: FYI @Aparna-Sakshi

Comment (fkiraly) [reply sktime/transformations/series/feature_selection.py:186]: (would be nice to finish the extension template PR 😃 )

Comment (fkiraly) [reply sktime/transformations/series/feature_selection.py:186]: besides, you can still use `_config` as before if it would be a blocker for this PR

Comment (aiwalter) [reply sktime/transformations/series/feature_selection.py]: yes will just use `_config` to get this PR done :)

Comment (fkiraly) [reply sktime/transformations/series/feature_selection.py]: Well, if your line in `_config` looks like this:
`FeatureSelectionTransformer: {"foo": "bar"}`

then the content of `get_test_params` should look like this:
`return {"foo": "bar"}`

Should be straightforward...

**Comment by fkiraly** [comment]
>Shouldn't this be implemented in the base class? The test parameters should then be a class variable similar to the tags.

@mloning, why should it be in the base class? The parameters will be different per concrete class.

>The test parameters should then be a class variable similar to the tags.

I disagree with the design, since that does not allow you to have code logic that determines multiple parameter settings, instead you have to write everything down manually.

**Comment by mloning** [comment]
> I disagree with the design, since that does not allow you to have code logic that determines multiple parameter settings, instead you have to write everything down manually.

Why, you can always make test params a list and iterate over it? Implementing the method with full docstring if you only return some values seems like a lot of duplication.

**Comment by fkiraly** [comment]
>Why, you can always make test params a list and iterate over it? Implementing the method with full docstring if you only return some values seems like a lot of duplication.

Sorry, I think you have not understood and/or I have not properly communicated what I mean.

I am referring to the case where it ismuch  easier not to paste the list into the method, as a list, but instead code that generates the list. Let's say you want to use a 5 x 5 x 5 test grid of parameters. Would you want to paste a list of length 125, typing it out, or write two lines of code that loop through the set?

Further, making it a method, you can inside call `get_test_params` or `create_test_instance` of any estimator components, thus modularizing your tests and ensuring isolation of test content to the class - e.g., if the default of the component changes, the default of the composite changes with it, ensuring that you can localize bugs to the class, which would not be the case in your solution.

**Review by mloning — APPROVED** [review]
Looks all good to me. We can update the test params in a separate PR I think.

**Comment by mloning** [comment]
> Sorry, I think you have not understood and/or I have not properly communicated what I mean.
> 
> I am referring to the case where it ismuch easier not to paste the list into the method, as a list, but instead code that generates the list. Let's say you want to use a 5 x 5 x 5 test grid of parameters. Would you want to paste a list of length 125, typing it out, or write two lines of code that loop through the set? And, which of the two options is easier to read, understand, and debug?
> 
> Further, making it a method, you can inside call `get_test_params` or `create_test_instance` of any estimator components, thus modularizing your tests and ensuring isolation of test content to the class - e.g., if the default of the component changes, the default of the composite changes with it, ensuring that you can localize bugs to the class, which would not be the case in your solution.

Is there a separate issue for this to discuss this in more detail?

**Comment by fkiraly** [comment]
>@fkiraly Is there a separate issue for this to discuss this in more detail?

Yes, of course:
https://github.com/alan-turing-institute/sktime/issues/1098

The generic issue was brought up on July 3, and the step exists since late August

**Review by fkiraly — APPROVED** [review]
Yes. looks good to me too now.
